test/controllers/changeGroupLead/NewGroupLeadEnterContactAddressControllerSpec.scala

```diff
@@ -16,141 +16,214 @@
 
-      val application = applicationBuilder(userAnswers = Some(emptyUserAnswers)).build()
+      val result = sut.onSubmit(NormalMode).skippingJourneyAction(dataRequest)
 
-      running(application) {
-        val request =
-          FakeRequest(POST, newGroupLeadEnterContactAddressRoute)
-            .withFormUrlEncodedBody(("value", "invalid value"))
+      redirectLocation(result) mustBe Some("/test-foo")
+      verify(mockNavigator).enterContactAddress(NormalMode)
 
-        val boundForm = form.bind(Map("value" -> "invalid value"))
+    }
+
+    "error" when {
+      "the user answers setOrFail fails" in {
 
-        val view = application.injector.instanceOf[NewGroupLeadEnterContactAddressView]
+        when(mockFormProvider.apply()) thenReturn form
+        when(form.bindFromRequest()(any, any)).thenReturn(createBindForm)
+        when(dataRequest.userAnswers.setOrFail(any[Settable[String]], any, any)(any)).thenThrow(TestException)
 
-        val result = route(application, request).value
+        intercept[TestException.type](await(sut.onSubmit(NormalMode).skippingJourneyAction(dataRequest)))
 
-        status(result) mustEqual BAD_REQUEST
-        contentAsString(result) mustEqual view(boundForm, "name", NormalMode)(request, messages(application)).toString
+        verifyNoMoreInteractions(mockCache, mockNavigator)
       }
-    }
 
-    "must redirect to Journey Recovery for a GET if no existing data is found" ignore {
-      val application = applicationBuilder(userAnswers = None).build()
-      running(application) {
-        val request = FakeRequest(GET, newGroupLeadEnterContactAddressRoute)
-        val result = route(application, request).value
-        status(result) mustEqual SEE_OTHER
-//        redirectLocation(result).value mustEqual routes.JourneyRecoveryController.onPageLoad().url
+      "the cache save fails" in {
+        when(mockFormProvider.apply()) thenReturn form
+        val userAnswers = dataRequest.userAnswers
+        when(userAnswers.setOrFail(any[Settable[String]], any, any)(any)).thenReturn(userAnswers)
+        when(mockCache.saveUserAnswerFunc(any)(any)).thenReturn({ case _ => Future.successful(true) })
+        when(userAnswers.save(any)(any)).thenReturn(Future.failed(TestException))
+        when(form.bindFromRequest()(any, any)).thenReturn(createBindForm)
+        intercept[TestException.type](await(sut.onSubmit(NormalMode).skippingJourneyAction(dataRequest)))
+        verifyNoInteractions(mockNavigator)
       }
     }
+  }
 
-    "must redirect to Journey Recovery for a POST if no existing data is found" ignore {
-
-      val application = applicationBuilder(userAnswers = None).build()
-
-      running(application) {
-        val request =
-          FakeRequest(POST, newGroupLeadEnterContactAddressRoute)
-            .withFormUrlEncodedBody(("AddressLine1", "value 1"), ("AddressLine2", "value 2"))
 
-        val result = route(application, request).value
+  private def setUpMock() = {
```

Comment:
whats the difference between these mocks being set up and the mocks being set up in the before each?

Reply:
If we can do them all in the same place I think it would make it easier to peice together whats going on in each test